src/main/java/org/usfirst/frc4048/common/swervev2/SwerveDrivetrain.java

```diff
@@ -0,0 +1,170 @@
+import edu.wpi.first.math.geometry.Rotation2d;
+import edu.wpi.first.math.geometry.Translation2d;
+import edu.wpi.first.math.kinematics.ChassisSpeeds;
+import edu.wpi.first.math.kinematics.SwerveDriveKinematics;
+import edu.wpi.first.math.kinematics.SwerveModuleState;
+import edu.wpi.first.math.trajectory.TrapezoidProfile;
+import edu.wpi.first.wpilibj2.command.SubsystemBase;
+import frc.robot.Constants;
+import frc.robot.swervev2.components.EncodedSwerveSparkMax;
+import frc.robot.swervev2.type.SparkMaxSwerveModule;
+
+
+public class SwerveDrivetrain extends SubsystemBase {
+    private final CANSparkMax frontLeftDrive;
+    private final CANSparkMax frontLeftTurn;
+    private final CANSparkMax backLeftDrive;
+    private final CANSparkMax backLeftTurn;
+    private final CANSparkMax frontRightDrive;
+    private final CANSparkMax frontRightTurn;
+    private final CANSparkMax backRightDrive;
+    private final CANSparkMax backRightTurn;
+
+    private final WPI_CANCoder frontLeftCanCoder;
+    private final WPI_CANCoder backLeftCanCoder;
+    private final WPI_CANCoder frontRightCanCoder;
+    private final WPI_CANCoder backRightCanCoder;
+
+
+    private final SparkMaxSwerveModule frontLeft;
+    private final SparkMaxSwerveModule frontRight;
+    private final SparkMaxSwerveModule backLeft;
+    private final SparkMaxSwerveModule backRight;
+
+    private final Translation2d frontLeftLocation = new Translation2d(Constants.ROBOT_LENGTH/2, Constants.ROBOT_WIDTH/2);
+    private final Translation2d frontRightLocation = new Translation2d(Constants.ROBOT_LENGTH/2, -Constants.ROBOT_WIDTH/2);
+    private final Translation2d backLeftLocation = new Translation2d(-Constants.ROBOT_LENGTH/2, Constants.ROBOT_WIDTH/2);
+    private final Translation2d backRightLocation = new Translation2d(-Constants.ROBOT_LENGTH/2, -Constants.ROBOT_WIDTH/2);
+    private final SwerveDriveKinematics kinematics = new SwerveDriveKinematics(frontLeftLocation,frontRightLocation,backLeftLocation,backRightLocation);
+
+    private double gyroOffset = 180;
+    private final AHRS navxGyro;
+    private double navxGyroValue;
+    private float filterRoll = 0;
+    
+
+    private double getGyro() {
+        return (navxGyro.getAngle() % 360)*-1; //ccw should be positive
+    }
+
+    public SwerveDrivetrain() {
+        navxGyro = new AHRS();
+        navxGyroValue = -1;
+
+        frontLeftDrive = new CANSparkMax(Constants.DRIVE_FRONT_LEFT_D, CANSparkMaxLowLevel.MotorType.kBrushless);
+        frontLeftTurn = new CANSparkMax(Constants.DRIVE_FRONT_LEFT_S, CANSparkMaxLowLevel.MotorType.kBrushless);
+        backLeftDrive = new CANSparkMax(Constants.DRIVE_BACK_LEFT_D, CANSparkMaxLowLevel.MotorType.kBrushless);
+        backLeftTurn = new CANSparkMax(Constants.DRIVE_BACK_LEFT_S, CANSparkMaxLowLevel.MotorType.kBrushless);
+        
+        frontRightDrive = new CANSparkMax(Constants.DRIVE_FRONT_RIGHT_D, CANSparkMaxLowLevel.MotorType.kBrushless);
```

Comment:
One way to try and arrange the configuration is to define a SwerveConfig class that takes the CAN ids, CANcoder, and then we just have to pass in 4 of those.

Reply:
done